compiler/src/iree/compiler/Codegen/LLVMGPU/TransformDialectExtensions/TransformDialectLLVMGPUExtensions.cpp

```diff
@@ -27,5 +35,118 @@ void mlir::iree_compiler::registerTransformDialectLLVMGPUExtension(
 // TODO: Maybe we need both a transform.iree.cpu.bufferize and a
 // transform.iree.gpu.bufferize rather than a single common bufferize op?
 
+//===---------------------------------------------------------------------===//
+// Patterns for ForeachThreadToGpu rewrite.
+//===---------------------------------------------------------------------===//
+
+struct ForeachThreadToGpuRewriter
+    : public OpRewritePattern<scf::ForeachThreadOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  FailureOr<SmallVector<Value>> returningMatchAndRewrite(
+      scf::ForeachThreadOp foreachThreadOp, PatternRewriter &rewriter) const;
+
+  LogicalResult matchAndRewrite(scf::ForeachThreadOp foreachThreadOp,
+                                PatternRewriter &rewriter) const override {
+    return returningMatchAndRewrite(foreachThreadOp, rewriter);
+  }
+};
+
+FailureOr<SmallVector<Value>>
+ForeachThreadToGpuRewriter::returningMatchAndRewrite(
+    scf::ForeachThreadOp foreachThreadOp, PatternRewriter &rewriter) const {
+  if (foreachThreadOp.getNumResults() > 0)
+    return foreachThreadOp->emitError(
+        "only bufferized scf.foreach_thread lowers to gpu.thread");
+  if (foreachThreadOp.getNumThreads().size() > 3)
+    return foreachThreadOp->emitError(
+        "scf.foreach_thread with rank > 3 does not lower to gpu.thread");
+
+  // Step 1. Create the gpu.thread ops
+  Location loc = foreachThreadOp.getLoc();
+  IndexType indexType = rewriter.getIndexType();
+  SmallVector<Value> threadCount = foreachThreadOp.getNumThreads();
+  SmallVector<gpu::Dimension, 3> gpuDims{gpu::Dimension::x, gpu::Dimension::y,
+                                         gpu::Dimension::z};
+  SmallVector<Value> threadOps;
+  threadOps.reserve(3);
+  for (int64_t idx : llvm::seq<int64_t>(0, threadCount.size())) {
+    threadOps.push_back(
+        rewriter.create<gpu::ThreadIdOp>(loc, indexType, gpuDims[idx]));
+  }
+  threadCount.resize(3, rewriter.create<arith::ConstantIndexOp>(loc, 1));
+
+  // Step 2. Move the body of foreachThreadOp after the op.
+  // Erase the terminator first, it will not be used.
+  rewriter.eraseOp(foreachThreadOp.getTerminator());
+  Block *block = foreachThreadOp->getBlock();
+  block->getOperations().splice(
+      Block::iterator(foreachThreadOp),
+      foreachThreadOp.getRegion().front().getOperations());
+
+  // Step 3. RAUW thread indices to thread ops.
+  for (auto it : llvm::zip(foreachThreadOp.getThreadIndices(), threadOps))
+    std::get<0>(it).replaceAllUsesWith(std::get<1>(it));
+
+  rewriter.eraseOp(foreachThreadOp);
+
+  return threadCount;
+}
+
+//===---------------------------------------------------------------------===//
+// IREE-specific LLVMGPU transformations.
+//===---------------------------------------------------------------------===//
+
+// TODO: if the number of threads was wired like the workgroup_count, we could
```

Comment:
Maybe I am mistaken, but I see direct mapping to threadids here?

Reply:
not sure how to parse this, is it related to the TODO ?
The TODO says we could have non-static dimensions too, although we prob often want the `tidx` to be static and a multiple of warp size (but I've seen many cases of it not being the best thing to do).

Reply:
I am trying to understand how the body of the `for_each` gets mapped to the hardware. Does the body map to a thread or a thread block? Seems like it is directly mapping to a thread?

Reply:
Yes, a foreach is mapped 1-1 to a threadblock for now and sets the threadblock.
It is the responsibility of the user of the pattern to make sure multiple such foreach_thread would be compatible within the same kernel.
Enhancements to this basic mode will come in the future on a per-need basis, should be pretty easy to extend.